packages/sanity/src/core/components/react-track-elements/hooks.ts

```diff
@@ -0,0 +1,116 @@
+  remove: (id: string) => void
+}
+
+function createStore<Value>(reportedValues: Map<string, Value>, publish: () => void) {
+  function add(id: string, value: Value) {
+    if (reportedValues.has(id)) {
+      // eslint-disable-next-line no-console
+      // console.error(
+      //   new Error(
+      //     `Invalid call to useReporter(${id}): A component reporting on "${id}" is already mounted in the subtree. Make sure that all reporters within the same <Tracker> subtree have unique ids.`
+      //   )
+      // )
+    }
+    reportedValues.set(id, value)
+    publish()
+  }
+
+  function update(id: string, value: Value) {
+    if (!reportedValues.has(id)) {
+      // throw new Error(`A reporter with id "${id}" is not known.`)
+    }
+    reportedValues.set(id, value)
+    publish()
+  }
+
+  function remove(id: string) {
+    if (!reportedValues.has(id)) {
+      // throw new Error(`A reporter with id "${id}" is not known`)
+    }
+    reportedValues.delete(id)
+    publish()
+  }
+
+  return {
+    add,
+    remove,
+    update,
+  }
+}
+
+/** @internal */
+export type TrackerContextGetSnapshot<Value> = [string, Value][]
+
+/** @internal */
+export function useTrackerStore<Value>(): {
+  store: TrackerContextStore<Value>
+  snapshot: TrackerContextGetSnapshot<Value>
+} {
+  const [reportedValues] = useState(() => new Map<string, Value>())
+  const [snapshot, updateSnapshot] = useReducer(() => Array.from(reportedValues.entries()), [])
+  const debouncedUpdateSnapshot = useMemo(() => debounce(updateSnapshot, 10, {trailing: true}), [])
+  const store = useMemo(
+    () => createStore(reportedValues, debouncedUpdateSnapshot),
+    [debouncedUpdateSnapshot, reportedValues],
+  )
+
+  return {store, snapshot}
+}
+
+/** @internal */
+export function useTrackerStoreReporter<Value>(
+  store: TrackerContextStore<Value> | null,
+  id: string | null,
+  value: () => Value,
+  isEqual: IsEqualFunction<Value> = Object.is,
+): void {
+  const idRef = useRef<string | null>(null)
+  const previousRef = useRef<Value | null>(null)
+
+  useLayoutEffect(() => {
+    /**
+     * Setup and teardown, only runs if `id`, `store` or the `value` getter changes
+     */
+    if (id === null || store === null) {
+      return undefined
+    }
+    const nextValue = value()
+    store.add(id, nextValue)
+    idRef.current = id
+    previousRef.current = nextValue
+    return () => {
+      store.remove(id)
+      idRef.current = null
+      previousRef.current = null
+    }
+  }, [id, store, value])
+
+  useLayoutEffect(() => {
+    /**
+     * Runs after every render in case the result of calling `value` has changes that
+     * the `isEqual` function picks up on.
+     * @TODO This is a bit expensive, and we should migrate to using a library like `@floating-ui/react` instead of rolling our own solution.
+     */
+    if (id === null || idRef.current === null || store === null || id !== idRef.current) {
+      return undefined
+    }
```

Comment:
I read your comment in the description but I have to admit that I'm still not quite sure how the "@floating/react" would help here.

It might be because I'm having a hard time placing what this is targetting but from what I understood it (can be) the presence circle with the avatar. I'm just wondering what's the benefit of this here. I get it very generically that it will improve the performance but I'm trying to understand why we should do these sort of checks here vs making sure that we use that sort of library within the @sanity/ui 

(I feel like this is a stupid question but I'm honestly curious so I'll just push through it 🥲)

Reply:
The comment is talking about the cost of running `isEqual` on `value` on every render, while pointing to `@floating-ui/react`, I get that's confusing since it's two things rolled into one.
Using `@floating-ui/react` to replace just this useEffect wouldn't do anything, but if we used `@floating-ui/react` to power both Presence Overlays, and Change Indicator connectors, then it has a whole pipeline built for doing what we're doing here, in a much more efficient way:
```
1. Register all DOM nodes. (useTrackerStoreReporter)
2. Calc position. (logic that uses the value of `useChangeIndicatorsReportedValues`, or `usePresenceReportedValues`, as input, and schedules a re-render as output)
3. Render with new position.
```
With `@floating-ui/react` we could build up both features with the built-ins for `computePosition()`, `autoPlacement()` and more, instead of rolling our own 🙌

Reply:
Might be something worth bringing to @runeb for a follow up? 👁️

Reply:
Sure 😄